### PR TITLE
[DUBBO-78] Add options configuration mechanism for JdkCompiler.

### PR DESCRIPTION
**Pull request by XenoAmess**
## What is the purpose of the change

I personally need this mechanism in one of my repos, (I have to copy/change this whole class now because not having such mechanism, I need a JdkCompiler to deal with java 1.8 codes)
and I think it helpful to other people too.

## Brief changelog

JdkCompiler

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by AlbumenJ** [comment]
Hi, @XenoAmess , you can customize a `AbstractCompiler` in your project using SPI.

**Review by AlbumenJ — APPROVED** [review]
LGTM.

**Comment by XenoAmess** [comment]
> Hi, @XenoAmess , you can customize a `AbstractCompiler` in your project using SPI.

Yep, now I copied the JdkCompiler and changed it to a Jdk8Compiler.
It can do.
see https://github.com/XenoAmess/metasploit-java-external-module/blob/develop/src/loader/src/main/java/com/xenoamess/metasploit/java_external_module/loader/Jdk8Compiler.java

I just think things can be easier... because that be just duplicated codes lol.

File: dubbo-common/src/test/java/org/apache/dubbo/common/compiler/support/JdkCompilerTest.java [hunk 1]
```diff
@@ -1,56 +1,87 @@
...
-        Assertions.assertThrows(IllegalStateException.class, () -> {
-            JdkCompiler compiler = new JdkCompiler();
-            Class<?> clazz = compiler.compile(getSimpleCodeWithSyntax(), JdkCompiler.class.getClassLoader());
-            Object instance = clazz.newInstance();
-            Method sayHello = instance.getClass().getMethod("sayHello");
-            Assertions.assertEquals("Hello world!", sayHello.invoke(instance));
-        });
-    }
-}
+/*
```
Comment (AlbumenJ) [inline dubbo-common/src/test/java/org/apache/dubbo/common/compiler/support/JdkCompilerTest.java:1]: please recover file format

Comment (XenoAmess) [reply dubbo-common/src/test/java/org/apache/dubbo/common/compiler/support/JdkCompilerTest.java]: @AlbumenJ done.

**Comment by XenoAmess** [comment]
@AlbumenJ 
still wondering...
why JdkCompiler.java  be CRLF but JdkCompilerTest.java be LF?
please consider do a unify...

**Comment by chickenlj** [comment]
Hi @XenoAmess Thank you for the contribution. We have reviewed and merged some of your pull requests recently. However, all the changes are code reformats which do not match the overall evolve plan of the community and will bring us a lot of conflicts and verifications.

I think it will be better for us to focus on such refactoring when we arrive at a certain stage. We have labeled all PR as `reopen later` so that we can review these PRs later.

Please feel free to let us know if you have any questions.

**Comment by XenoAmess** [comment]
> all the changes are code reformats

Not all of the prs are code reformats. for example, this one is not.

I wonder, have you at least read them all?

I will reopen the prs which be not reformat pr, and I hope you can at least find some time for them.

> I think it will be better for us to focus on such refactoring when we arrive at a certain stage. 

According to my former experience there will never be such a perfect stage.

But if you wanna delay some of them, I have no problem, as this is not my lib to maintain.

> which do not match the overall evolve plan of the community

Well, you have different ideas about this with people in alibaba/p3c. As you are both alibaba team, I'm actually confused.

I thought you alibaba people will follow one same idea about code maintaining, and I heard every java repo in ali must pass p3c. Maybe I be wrong.

**Comment by XenoAmess** [comment]
@chickenlj this pr is a functional pr, should not be closed.

**Comment by AlbumenJ** [comment]
@XenoAmess 

In my opinion, this pr should be reopened.

Why we temporarily close those reformatting PRs is because we are currently resolve most of previous PRs which created for servals months and if we merge those reformatting PRs  may cause a lot of confilcts with the previous PRs .

Code maintaining is important, and we will make code clean enough once after we resolve those previous PRs.

We are sure that will reopen those reformatting PRs in the future, Also please feel free to notify us if we forget it.

**Review by AlbumenJ — APPROVED** [review]
LGTM.

**Comment by XenoAmess** [comment]
> Also, we cannot find if the file is in CRLF mode only from Github Pull Request website and it is sure that the files in the future PR in CRLF mode is unavoidable.

This can be done using checkstyle plugin.

We can fail the build when there be CRLF file.

See example at commons-lang.

https://github.com/apache/commons-lang/blob/master/pom.xml#L754

https://github.com/apache/commons-lang/blob/master/src/site/resources/checkstyle/checkstyle.xml#L28

File: dubbo-common/src/main/java/org/apache/dubbo/common/compiler/support/JdkCompiler.java [hunk 2]
```diff
@@ -61,14 +61,22 @@
...
-    private volatile List<String> options;
+    private final List<String> options;
 
-    public JdkCompiler() {
-        options = new ArrayList<String>();
-        options.add("-source");
-        options.add("1.6");
-        options.add("-target");
-        options.add("1.6");
+    private static final String DEFAULT_JAVA_VERSION = "1.6";
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/compiler/support/JdkCompiler.java:66]: shall we upgrade this to 1.8 ?

Comment (XenoAmess) [reply dubbo-common/src/main/java/org/apache/dubbo/common/compiler/support/JdkCompiler.java:66]: @AlbumenJ 

Well in my own usecases I need 1.8, so I want such a mechanism to configure it.

However I really do not know why you originally chose 1.6 here.

Is there any special reason?

I myself will not stop you from change it to 1.8, but I really don't know about other users of this class's opinions.

Comment (AlbumenJ) [reply dubbo-common/src/main/java/org/apache/dubbo/common/compiler/support/JdkCompiler.java:66]: For Dubbo 2.7.x, jdk 1.6 is not compatiable.

Comment (XenoAmess) [reply dubbo-common/src/main/java/org/apache/dubbo/common/compiler/support/JdkCompiler.java:66]: > For Dubbo 2.7.x, jdk 1.6 is not compatiable.

@AlbumenJ sounds reasonable.
OK, then we change it to 1.8.

**Review by AlbumenJ — APPROVED** [review]
LGTM.

**Pull request merged**